### PR TITLE
[llvm][SLPVectorizer] Fix a bad cast assertion

### PR DESCRIPTION
**Pull request by jroelofs**

### CODE REVIEW

**Comment by jroelofs** [comment]
> @alexey-bataev changed the title

FWIW, the Developer Policy still [allows](https://llvm.org/docs/DeveloperPolicy.html#commit-messages) such links in commit messages.

**Review by alexey-bataev — APPROVED** [review]
LG

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -361,12 +361,14 @@ static bool isCommutative(Instruction *I) {
   return I->isCommutative();
 }
 
-/// \returns inserting index of InsertElement or InsertValue instruction,
-/// using Offset as base offset for index.
-static std::optional<unsigned> getInsertIndex(const Value *InsertInst,
-                                              unsigned Offset = 0) {
+template <typename T>
+static std::optional<unsigned> getInsertExtractIndex(const Value *Inst,
```
Comment (inclyc) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:365]:
```suggestion
std::optional<unsigned> getInsertExtractIndex(const Value *Inst,
```

A little strange with that `static`, maybe better declares all these helpers in anonymous namespace?

**Review by inclyc — APPROVED** [review]

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 1 above)
Comment (jroelofs) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:365]: eh, not that weird:
```
$ git grep "template <.*> static" llvm | wc -l
     132
```

Comment (inclyc) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:365]: Yes I think you can commit this directly if you like, I don't have strong opinion about this.

**Pull request merged**